proofs/erc20.md

```diff
@@ -0,0 +1,1030 @@
+
+```
+#hashedLocation({lang}, pos(map), key1 key2)
+```
+
+This abstraction makes the specification independent of the underlying compilers, enabling it to be reused for differently compiled programs.
+
+
+### Gas Consumption
+
+Regarding the gas consumption of the program, the maximum gas amount is specified to ensure that the program does not consume more gas than the limit.
+
+The verifier proves that the actual gas consumption is less than the provided limit, and also reports the exact amount of gas consumed during the execution. Indeed, it reports a set of the amounts since the gas consumption varies depending on the context (i.e., the input parameter values and the state of the storage). This provides the most precise gas analysis.
+
+
+
+We present the specification for each function defined in the ERC20 standard. Here we focus on explaining the EVM-specific detailed behaviors, referring to the ERC20-K specification for the high-level logic.
+
+
+totalSupply
+-----------
+
+```{.k .erc20-spec .hkg-spec}
+[totalSupply]
+```
+
+`<k>` cell specifies that the execution eventually reaches RETURN instruction meaning that the program will successfully terminate. The RETURN instruction says that a 32-byte return value will be stored in the memory at the location RET_ADDR.
+
+```{.k .erc20-spec .hkg-spec}
+k: #execute => (RETURN RET_ADDR:Int 32 ~> _)
+
+callData: #abiCallData("totalSupply", .TypedArgs)
+```
+
+`<localMem>` cell specifies that the local memory is empty in the beginning, but in the end, it will contain some contents including the return value `TOTAL`, the total supply of the tokens.
+
+The other entries are represented by the wildcard symbol `_`, meaning that they are not relevant to the functional correctness of the program.
+
+```{.k .erc20-spec .hkg-spec}
+localMem:
+      .Map
+    =>
+    (
+      .Map[ RET_ADDR := #asByteStackInWidth(TOTAL, 32) ]
+      _:Map
+    )
+```
+
+`<gas>` cell specifies the maximum gas amount. Here we give a loose upper-bound for the demonstration purpose. In practice, one should set a reasonable amount of the gas limit to see if the program does not consume too much gas (i.e., no gas leakage).
+
+```{.k .erc20-spec .hkg-spec}
+gas: {gascap} => _
+```
+
+`<log>` cell specifies that no log is generated during the execution.
+
+```{.k .erc20-spec .hkg-spec}
+log: _
+```
+
+`<refund>` cell specifies that no gas is refunded. Note that it does not mean it consumes all the provided gas. The gas refund is different from returning the remaining gas after the execution. It is another notion to capture some specific gas refund events that happen, for example, when an unused storage entry is re-claimed (i.e., garbage-collected). The following specification ensures that no such event happens during the execution of the current function.
+
+```{.k .erc20-spec .hkg-spec}
+refund: _
+```
```

Comment:
It would be nice to not have to specify the cells that do not change at all!  Narrowing should be able to do that for us.